### PR TITLE
Add working serial passthru functionality for flashing ELRS recievers

### PR DESCRIPTION
**Pull request by countofkrakow**
Courtesy of [Capn Bry](https://github.com/CapnBry/CRServoF/issues/29)

### CODE REVIEW

**Comment by mikeneiderhauser** [comment]
Will hopefully be looking at this PR this weekend time permitting

**Comment by L1cardo** [comment]
@countofkrakow I love you bro! Thanks for the efforts!

**Comment by L1cardo** [comment]
@countofkrakow Any tips for flashing more corroctly?
I can flash through CRSFJoystick 2.0 dongle 8/10, but when I switched to wareshare-rp2040, I keep getting these two errors.

image
image

**Comment by countofkrakow** [comment]
No idea, haven't tested on the waveshare. Only the Xiao seed and your pcb dongle.

**Comment by L1cardo** [comment]
So we can confirm now:
Works fine: 
- Xiao seed
- CRSFJoystick2.0 dongle

Don't know for sure:
- Wareshare-rp2040 zero


Let's wait for @mikeneiderhauser  if it works on his side.

**Comment by mikeneiderhauser** [comment]
Passthrough flashing of the waveshare rp2040 board with a happymodel EP2 as the RX went first time for me without issue.  I will be hitting a few comments on the PR. Once addressed I think we can land this!  Thanks for the contribution.

Tested flight in liftoff post passthrough flash. I did not notice any perf issues.

Passthrough flash log below.


`init_passthrough(["upload"],` [".pio\build\HappyModel_EP_2400_RX_via_BetaflightPassthrough\firmware.bin"])
Using manually specified: COM3
======== PASSTHROUGH INIT ========
  Trying to initialize COM3 @ 420000

Attempting to detect FC UART configuration...
    ** Serial RX config detected: 'serial 5 64 0 0 0 0'
Enabling serial passthrough...
  CMD: 'serialpassthrough 5 420000'
======== PASSTHROUGH DONE ========
======== RESET TO BOOTLOADER ========
  * Using full duplex (CRSF)
Verified RX target 'HAPPYMODEL_EP_2400_RX'
Configuring upload protocol...
AVAILABLE: espota, esptool
CURRENT: upload_protocol = esptool
Looking for upload port...
Using manually specified: COM3
Uploading .pio\build\HappyModel_EP_2400_RX_via_BetaflightPassthrough\firmware.bin
esptool.py v4.2.1
Serial port COM3
WARNING: Pre-connection option "no_reset" was selected. Connection may fail if the chip is not in bootloader or flasher stub mode.
Connecting....
Chip is ESP8285H16
Features: WiFi, Embedded Flash
WARNING: Detected crystal freq 41.37MHz is quite different to normalized freq 40MHz. Unsupported crystal in use?
Crystal is 40MHz
MAC: 4c:eb:d6:10:77:4e
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00000000 to 0x0007afff...
Compressed 502704 bytes to 367992...
Writing at 0x00000000... (0 %)
Writing at 0x0000094a... (1 %)
Writing at 0x00001624... (1 %)
Writing at 0x00002087... (2 %)
Writing at 0x00002bda... (2 %)
Writing at 0x000037ed... (3 %)
Writing at 0x00004525... (3 %)
Writing at 0x00005013... (4 %)
Writing at 0x000061c5... (5 %)
Writing at 0x00006cf2... (5 %)
Writing at 0x0000777c... (6 %)
Writing at 0x000083b1... (6 %)
Writing at 0x00008f18... (7 %)
Writing at 0x00009a43... (7 %)
Writing at 0x0000a5c6... (8 %)
Writing at 0x0000b16d... (8 %)
Writing at 0x0000bddf... (9 %)
Writing at 0x0000c8cd... (10 %)
Writing at 0x0000d4c4... (10 %)
Writing at 0x0000e340... (11 %)
Writing at 0x0000ee96... (11 %)
Writing at 0x0000f9f8... (12 %)
Writing at 0x00010732... (12 %)
Writing at 0x0001136e... (13 %)
Writing at 0x00011eaf... (13 %)
Writing at 0x00012bc8... (14 %)
Writing at 0x00013869... (15 %)
Writing at 0x000142f0... (15 %)
Writing at 0x00014e56... (16 %)
Writing at 0x000158a3... (16 %)
Writing at 0x000163d0... (17 %)
Writing at 0x00016e9e... (17 %)
Writing at 0x00017977... (18 %)
Writing at 0x00018478... (18 %)
Writing at 0x0001905b... (19 %)
Writing at 0x00019b3c... (20 %)
Writing at 0x0001a620... (20 %)
Writing at 0x0001b2d0... (21 %)
Writing at 0x0001bce1... (21 %)
Writing at 0x0001c803... (22 %)
Writing at 0x0001d39c... (22 %)
Writing at 0x0001e020... (23 %)
Writing at 0x0001eb80... (23 %)
Writing at 0x0001f807... (24 %)
Writing at 0x00020295... (25 %)
Writing at 0x00020e06... (25 %)
Writing at 0x000219f9... (26 %)
Writing at 0x00022544... (26 %)
Writing at 0x000230ee... (27 %)
Writing at 0x00023bcd... (27 %)
Writing at 0x0002472f... (28 %)
Writing at 0x00025743... (28 %)
Writing at 0x00026322... (29 %)
Writing at 0x00026f94... (30 %)
Writing at 0x00027b5e... (30 %)
Writing at 0x00028722... (31 %)
Writing at 0x0002960e... (31 %)
Writing at 0x0002a2c1... (32 %)
Writing at 0x0002b12d... (32 %)
Writing at 0x0002bd0e... (33 %)
Writing at 0x0002c889... (33 %)
Writing at 0x0002d3f0... (34 %)
Writing at 0x0002de4d... (35 %)
Writing at 0x0002ea18... (35 %)
Writing at 0x0002f486... (36 %)
Writing at 0x0002ff20... (36 %)
Writing at 0x00030903... (37 %)
Writing at 0x00031271... (37 %)
Writing at 0x00031f06... (38 %)
Writing at 0x00032a25... (38 %)
Writing at 0x0003362d... (39 %)
Writing at 0x000340be... (40 %)
Writing at 0x00034ba1... (40 %)
Writing at 0x00035636... (41 %)
Writing at 0x000360e7... (41 %)
Writing at 0x00036b89... (42 %)
Writing at 0x000376c1... (42 %)
Writing at 0x0003817e... (43 %)
Writing at 0x00038d67... (43 %)
Writing at 0x000397d6... (44 %)
Writing at 0x0003a2b4... (45 %)
Writing at 0x0003ae88... (45 %)
Writing at 0x0003b9cc... (46 %)
Writing at 0x0003c543... (46 %)
Writing at 0x0003cfc9... (47 %)
Writing at 0x0003d9f9... (47 %)
Writing at 0x0003e48c... (48 %)
Writing at 0x0003ee42... (48 %)
Writing at 0x0003f7fb... (49 %)
Writing at 0x000401a8... (50 %)
Writing at 0x00040b8d... (50 %)
Writing at 0x00041621... (51 %)
Writing at 0x00042039... (51 %)
Writing at 0x000429ee... (52 %)
Writing at 0x000432f2... (52 %)
Writing at 0x00043cfe... (53 %)
Writing at 0x00044772... (53 %)
Writing at 0x00045170... (54 %)
Writing at 0x00045b56... (55 %)
Writing at 0x000465ad... (55 %)
Writing at 0x00047015... (56 %)
Writing at 0x00047a3e... (56 %)
Writing at 0x0004848a... (57 %)
Writing at 0x00048e25... (57 %)
Writing at 0x00049852... (58 %)
Writing at 0x0004a2af... (58 %)
Writing at 0x0004ad14... (59 %)
Writing at 0x0004b741... (60 %)
Writing at 0x0004c1a6... (60 %)
Writing at 0x0004cdfb... (61 %)
Writing at 0x0004d906... (61 %)
Writing at 0x0004e4d3... (62 %)
Writing at 0x0004f020... (62 %)
Writing at 0x0004fb45... (63 %)
Writing at 0x000505c7... (63 %)
Writing at 0x0005102b... (64 %)
Writing at 0x00051af4... (65 %)
Writing at 0x000526b9... (65 %)
Writing at 0x00053114... (66 %)
Writing at 0x00053c36... (66 %)
Writing at 0x000547a5... (67 %)
Writing at 0x00055192... (67 %)
Writing at 0x00055d3f... (68 %)
Writing at 0x00056793... (68 %)
Writing at 0x000571c0... (69 %)
Writing at 0x00057caa... (70 %)
Writing at 0x000587b9... (70 %)
Writing at 0x000591e6... (71 %)
Writing at 0x00059bc7... (71 %)
Writing at 0x0005a5c3... (72 %)
Writing at 0x0005b0b6... (72 %)
Writing at 0x0005bb77... (73 %)
Writing at 0x0005c5de... (73 %)
Writing at 0x0005d000... (74 %)
Writing at 0x0005db24... (75 %)
Writing at 0x0005e675... (75 %)
Writing at 0x0005f18c... (76 %)
Writing at 0x0005fcab... (76 %)
Writing at 0x0006075b... (77 %)
Writing at 0x00061249... (77 %)
Writing at 0x00061a8b... (78 %)
Writing at 0x0006228a... (78 %)
Writing at 0x00062a8d... (79 %)
Writing at 0x00063293... (80 %)
Writing at 0x00063a90... (80 %)
Writing at 0x0006428e... (81 %)
Writing at 0x00064a8a... (81 %)
Writing at 0x00065263... (82 %)
Writing at 0x00065a62... (82 %)
Writing at 0x00066293... (83 %)
Writing at 0x00066a93... (83 %)
Writing at 0x000672a3... (84 %)
Writing at 0x00067aa2... (85 %)
Writing at 0x000682a1... (85 %)
Writing at 0x00068aa1... (86 %)
Writing at 0x00069246... (86 %)
Writing at 0x00069a0f... (87 %)
Writing at 0x0006a1d0... (87 %)
Writing at 0x0006a9a2... (88 %)
Writing at 0x0006b176... (88 %)
Writing at 0x0006b93d... (89 %)
Writing at 0x0006c7d0... (90 %)
Writing at 0x0006dbf4... (90 %)
Writing at 0x0006e95e... (91 %)
Writing at 0x0006f2ce... (91 %)
Writing at 0x0006ff8a... (92 %)
Writing at 0x00070a79... (92 %)
Writing at 0x00071668... (93 %)
Writing at 0x0007212f... (93 %)
Writing at 0x00072b74... (94 %)
Writing at 0x000735b5... (95 %)
Writing at 0x00073f9e... (95 %)
Writing at 0x00074a25... (96 %)
Writing at 0x00075385... (96 %)
Writing at 0x00075dff... (97 %)
Writing at 0x00076876... (97 %)
Writing at 0x000773ac... (98 %)
Writing at 0x0007829c... (98 %)
Writing at 0x00079290... (99 %)
Writing at 0x0007a005... (100 %)
Wrote 502704 bytes (367992 compressed) at 0x00000000 in 20.0 seconds (effective 201.2 kbit/s)...
Hash of data verified.

Leaving...
Soft resetting...
========================= [SUCCESS] Took 36.89 seconds =========================

Environment                                      Status    Duration
-----------------------------------------------  --------  ------------
HappyModel_EP_2400_RX_via_BetaflightPassthrough  SUCCESS   00:00:36.893
========================= 1 succeeded in 00:00:36.893 =========================`

File: Sketchbook/CRSF_Joystick/CRSF_Joystick.ino [hunk 1]
```diff
@@ -4,16 +4,20 @@
 #include "calibration.h"
 
 // ONLY UNCOMMENT 1 Board ID
-//#define BOARD_ID 1  // PiPico
+#define BOARD_ID 1  // PiPico
 //#define BOARD_ID 2  // TINY2040
-#define BOARD_ID 3  // WaveShare RP2040-Zero
+//#define BOARD_ID 3  // WaveShare RP2040-Zero
```
Comment (mikeneiderhauser) [inline Sketchbook/CRSF_Joystick/CRSF_Joystick.ino:9]: Need to make the WaveShare the default board.  I have boards in the field that individuals will want to flash.

**Comment by countofkrakow** [comment]
Why not?

**Comment by L1cardo** [comment]
> Why not?

Don't know, I always fail when it is connected.

**Comment by howels** [comment]
> Why not?

Should not make a difference with eLRS RX at all, the RX will enter bootloader when requested and link to the TX once flash is completed.

**Comment by L1cardo** [comment]
Always get this issue when RX is connected.

image

File: Sketchbook/CRSF_Joystick/CRSF_Joystick.ino [hunk 2]
```diff
@@ -4,16 +4,20 @@
 #include "calibration.h"
 
 // ONLY UNCOMMENT 1 Board ID
-//#define BOARD_ID 1  // PiPico
+#define BOARD_ID 1  // PiPico
```
Comment (mikeneiderhauser) [inline Sketchbook/CRSF_Joystick/CRSF_Joystick.ino:7]: Need to comment out board id 1.  The define below will technically override it, but worth keeping it clean.

**Pull request merged**